### ISSUE TITLE
Rocket.Chat server crashed due to startRocketChat program being killed

### ISSUE THREAD

**Issue by sst2010** [RocketChat/Rocket.Chat#17634]
Hey there. I'm currently hosting the `Rocket.Chat` server at https://chat.incendia.in

### Server Specs:
- OS: Ubuntu 18.04
- Container: snap
- RAM: 2 GB
- Swap: 2 GB

### Problem:

This morning, I woke up to find that the server had crashed in the middle of the night.

- I logged in to the server & attempted to restart the `rocketchat-server` with
  ```
  sudo service snap.rocketchat-server.rocketchat-server restart
  ```
  - The server restart failed with error stating: `mongo cannot be connected, ensure its running`
    - Logs:
      ```
      May 13 00:45:10 systemd[1]: snap.rocketchat-server.rocketchat-server.service: Service hold-off time over, scheduling restart.
      May 13 00:45:10 systemd[1]: snap.rocketchat-server.rocketchat-server.service: Scheduled restart job, restart counter is at 212.
      May 13 00:45:10 systemd[1]: Stopped Service for snap application rocketchat-server.rocketchat-server.
      May 13 00:45:10 systemd[1]: Started Service for snap application rocketchat-server.rocketchat-server.
      May 13 00:45:10 rocketchat-server.rocketchat-server[19843]: Mongo is not available, can't start. Waiting 10 seconds and trying again
      May 13 00:45:20 rocketchat-server.rocketchat-server[19843]: Mongo is not available, can't start. Waiting 15 seconds and trying again
      May 13 00:45:35 rocketchat-server.rocketchat-server[19843]: Mongo is not available, can't start. Waiting 20 seconds and trying again
      ```

- I checked that Mongo was not running with `sudo netstat -plnt`. So, I tried to restart `rocketchat-mongo` with
  ```
  sudo service snap.rocketchat-server.rocketchat-mongo restart
  ```
  - Mongo restart failed with error stating: `mongo cannot be connected, ensure its running`
    - Logs:
      ```
      May 13 00:46:27 systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Main process exited, code=exited, status=100/n/a
      May 13 00:46:27 systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Failed with result 'exit-code'.
      May 13 00:46:27 systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Service hold-off time over, scheduling restart.
      May 13 00:46:27 systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Scheduled restart job, restart counter is at 5.
      May 13 00:46:27 systemd[1]: Stopped Service for snap application rocketchat-server.rocketchat-mongo.
      May 13 00:46:27 systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Start request repeated too quickly.
      May 13 00:46:27 systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Failed with result 'exit-code'.
      May 13 00:46:27 systemd[1]: Failed to start Service for snap application rocketchat-server.rocketchat-mongo.
      ```

- Googling the issue led me here: https://forums.rocket.chat/t/mongo-is-not-available-cant-start/382
  - The suggested solution was to ensure that I do not have multiple installations of mongo.
  - A search of installed packages with `sudo dpkg -l | grep mongo` revealed I do have `mongodb-server` installed.
  - Since, I had installed Rocket.Chat with `snap`, I wasn't sure whether this should be here or not.
  - Also, uninstalling this `mongodb-server` seemed risky ( I'm not sure but it might result in further issues or loss of data. ).

  
- So, I simply restarted the server & everything started perfectly.

- I digged into the logs to identify the crash point to figure out the reason for this crash.
- The server seemed to crash due to a `startRocketChat` program being killed.
- Please, see logs below:
  ```
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]: UserPresence:away { id: 'J5aCy9D3FoMebQPdz',
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:   clientAddress: '157.32.140.99',
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:   httpHeaders:
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:    { referer: 'https://chat.incendia.in/group/8gmd4sarzyxu7p4zx',
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:      'x-forwarded-for': '157.32.140.99',
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:      'x-real-ip': '157.32.140.99',
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:      host: 'chat.incendia.in',
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:      'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Rocket.Chat/2.17.7 Chrome/7
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:      'accept-language': 'en-US' },
  May 12 17:44:38 rocketchat-server.rocketchat-server[6643]:   userId: 'pcaDYqyuqN69R5eSz' }
  May 12 18:49:55 rocketchat-server.rocketchat-server[6643]: /snap/rocketchat-server/1427/bin/startRocketChat: line 3:  6732 Killed                  node $SNAP/main.js
  May 12 18:49:55 systemd[1]: snap.rocketchat-server.rocketchat-server.service: Main process exited, code=exited, status=137/n/a
  May 12 18:49:55 systemd[1]: snap.rocketchat-server.rocketchat-server.service: Failed with result 'exit-code'.
  May 12 18:49:56 systemd[1]: snap.rocketchat-server.rocketchat-server.service: Service hold-off time over, scheduling restart.
  May 12 18:49:56 systemd[1]: snap.rocketchat-server.rocketchat-server.service: Scheduled restart job, restart counter is at 8.
  May 12 18:49:56 systemd[1]: Stopped Service for snap application rocketchat-server.rocketchat-server.
  May 12 18:49:56 systemd[1]: Started Service for snap application rocketchat-server.rocketchat-server.
  May 12 18:50:03 systemd[1]: Started Service for snap application rocketchat-server.rocketchat-server.
  May 12 18:50:04 rocketchat-server.rocketchat-server[7184]: Mongo is not available, can't start. Waiting 10 seconds and trying again
  May 12 18:50:14 rocketchat-server.rocketchat-server[7184]: Mongo is not available, can't start. Waiting 15 seconds and trying again
  May 12 18:50:29 rocketchat-server.rocketchat-server[7184]: Mongo is not available, can't start. Waiting 20 seconds and trying again
  May 12 18:50:49 rocketchat-server.rocketchat-server[7184]: Mongo is not available, can't start. Waiting 25 seconds and trying again
  May 12 18:51:14 rocketchat-server.rocketchat-server[7184]: Mongo is not available, can't start. Waiting 30 seconds and trying again
  ```

The server is running successfully after the reboot. But, I would like to know why this happened & what I can do to prevent it. If its a bug, the development team can take it up as well.
- Note that the server has been running successfully for almost 3 weeks.

Please let me know if any more information is needed from my end. Many thanks in advance.

**Comment by sisomoti** [comment]
We have similar problems since today.
The rocketchat-server.rocketchat-server is active, but the rocketchat-server.rocketchat-mongo is inactive.

We found below output in /var/log/syslog ,so run journalctl command.
```
sudo systemctl status snap.rocketchat-server.rocketchat-mongo to view logs: sudo journalctl -u snap.rocketchat-server.rocketchat-mongo
```

logs here:
```
sudo journalctl -u snap.rocketchat-server.rocketchat-mongo | less


-- Logs begin at 水 2020-05-13 11:39:47 JST, end at 水 2020-05-13 12:32:43 JST. --
systemd[1]: Started Service for snap application rocketchat-server.rocketchat-mongo.
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] MongoDB starting : pid=1092 port=27017 dbpath=/var/snap/rocketchat-server/common 64-bit host=rocketchat-my-host
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] db version v3.4.20
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] git version: 447847d93d6e0a21b018d5df45528e815c7c13d8
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] allocator: tcmalloc
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] modules: none
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] build environment:
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten]     distarch: x86_64
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten]     target_arch: x86_64
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] options: { net: { bindIp: "127.0.0.1" }, processManagement: { pidFilePath: "/var/snap/rocketchat-server/common/mongod.pid" }, replication: { replSet: "rs0" }, storage: { dbPath: "/var/snap/rocketchat-server/common", journal: { enabled: true }, mmapv1: { smallFiles: true } } }
rocketchat-server.rocketchat-mongo[882]:  I -        [initandlisten] Detected data files in /var/snap/rocketchat-server/common created by the 'wiredTiger' storage engine, so setting the active storage engine to 'wiredTiger'.
rocketchat-server.rocketchat-mongo[882]:  I STORAGE  [initandlisten]
rocketchat-server.rocketchat-mongo[882]:  I STORAGE  [initandlisten] ** WARNING: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine
rocketchat-server.rocketchat-mongo[882]:  I STORAGE  [initandlisten] **          See http://dochub.mongodb.org/core/prodnotes-filesystem
rocketchat-server.rocketchat-mongo[882]:  I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=1460M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),checkpoint=(wait=60,log_size=2GB),statistics_log=(wait=0),verbose=(recovery_progress),
rocketchat-server.rocketchat-mongo[882]:  E STORAGE  [initandlisten] WiredTiger error (-31802) [1589337600:179208][1092:0x7f3e53ae4d00], txn-recover: unsupported WiredTiger file version: this build  only supports major/minor versions up to 1/0,  and the file is version 2/0: WT_ERROR: non-specific WiredTiger error
rocketchat-server.rocketchat-mongo[882]:  E STORAGE  [initandlisten] WiredTiger error (0) [1589337600:179283][1092:0x7f3e53ae4d00], txn-recover: WiredTiger is unable to read the recovery log.
rocketchat-server.rocketchat-mongo[882]:  E STORAGE  [initandlisten] WiredTiger error (0) [1589337600:179302][1092:0x7f3e53ae4d00], txn-recover: This may be due to the log files being encrypted, being from an older version or due to corruption on disk
rocketchat-server.rocketchat-mongo[882]:  E STORAGE  [initandlisten] WiredTiger error (0) [1589337600:179318][1092:0x7f3e53ae4d00], txn-recover: You should confirm that you have opened the database with the correct options including all encryption and compression options
rocketchat-server.rocketchat-mongo[882]:  E STORAGE  [initandlisten] WiredTiger error (-31802) [1589337600:179340][1092:0x7f3e53ae4d00], txn-recover: Recovery failed: WT_ERROR: non-specific WiredTiger error
rocketchat-server.rocketchat-mongo[882]:  I -        [initandlisten] Assertion: 28595:-31802: WT_ERROR: non-specific WiredTiger error src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp 277
rocketchat-server.rocketchat-mongo[882]:  I STORAGE  [initandlisten] exception in initAndListen: 28595 -31802: WT_ERROR: non-specific WiredTiger error, terminating
rocketchat-server.rocketchat-mongo[882]:  I NETWORK  [initandlisten] shutdown: going to close listening sockets...
rocketchat-server.rocketchat-mongo[882]:  I NETWORK  [initandlisten] removing socket file: /tmp/mongodb-27017.sock
rocketchat-server.rocketchat-mongo[882]:  I NETWORK  [initandlisten] shutdown: going to flush diaglog...
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] now exiting
rocketchat-server.rocketchat-mongo[882]:  I CONTROL  [initandlisten] shutting down with code:100
systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Main process exited, code=exited, status=100/n/a
systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Unit entered failed state.
systemd[1]: snap.rocketchat-server.rocketchat-mongo.service: Failed with result 'exit-code'.
```

I found suspicious message in this logs...
```
WiredTiger error (-31802) [1589337600:179208][1092:0x7f3e53ae4d00], txn-recover:
unsupported WiredTiger file version:
this build  only supports major/minor versions up to 1/0,  and the file is version 2/0:
WT_ERROR: non-specific WiredTiger error
```

I don't know why occure this error & how to avoid/fix this problem, but this is maybe a bug of the mongoDB of the Rocket.Chat.

---

Related issue(maybe):
https://forums.rocket.chat/t/mongo-is-not-available-cant-start-waiting-30-seconds-and-trying-again/6930
or
https://forums.rocket.chat/t/sudden-502-error/6914/2

Since today version there are various symptopms, but there looks like that they having problems with the mongoDB equally.


Thanks for reading my bad damn English post.

**Comment by sisomoti** [comment]
issue #17628 is the same problem?
sorry if unrelated problem.

**Comment by sst2010** [comment]
@sisomoti 

You're right it does look like a problem similar to the one you mentioned.

Running `snap changes` gives the output:
```
ID   Status  Spawn                   Ready                   Summary
80   Done    yesterday at 06:41 UTC  yesterday at 06:41 UTC  Running service command
81   Done    yesterday at 18:45 UTC  yesterday at 18:50 UTC  Auto-refresh snap "rocketchat-server"
82   Done    today at 01:02 UTC      today at 01:02 UTC      Auto-refresh snap "rocketchat-server"
```

Looks like the server crashed 4 minutes after the snap was updated. So, this is definitely an issue with the update.

I'm turning off auto updates for now to avoid abrupt crashes in the future. To do do, run
```
sudo snap set system refresh.hold="2500-01-01T00:00:00+00:00"
```

I know auto-refresh is good. But, for some people, avoiding abrupt crashes might be more important. I personally do not mind manually updating periodically.
- To update snaps manually, run:
  ```
  sudo snap refresh
  ```

**Comment by sst2010** [comment]
Closing the issue as the major discussion seems to already be happening here: https://github.com/RocketChat/Rocket.Chat/issues/17628

**Comment by sisomoti** [comment]
@sst2010 thanks for information!

I have updated snaps manually(snap refresh), then rocketchat-mongo has being active!
note: after update, my rocketchat-server version is 2.4.11/ rocketchat-mong version is 3.6.14.


> I know auto-refresh is good. But, for some people, avoiding abrupt crashes might be more important.

Yes! YES!!!!! I had got crash after updating server so many times, and I feeling annoyed about this feature each time.
But according to your inform, it be able to hold stable version!!! (I know this for the first time)

Now  I don't have to worry about server update!
I really very very appreciate!!!
